sql: use timestamps for consistent, retryable multi-txn backups

madelynnblue opened the pull request:
Using logic similar to AS OF SYSTEM TIME, allow the creation of multiple
transactions that will remain consistent because their timestamps cannot
advance. Enable retries of each operation. In the future the fetch loop
could be made to work in parallel, but would need some mechanism to
limit the number of reads at once.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8819)

<!-- Reviewable:end -->

paperstreet commented:
Awesome! Have you tried running this on the registration cluster data?

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-KQ0zYdMdXo5CdK-G3bB:-KQ0zYdMdXo5CdK-G3bC:b-evprii) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d71e3438212c1fbb9143a580fbf20bf8cef243c/sql/backup.go#L87)):*
> ```Go
> //
> // TODO(dan): Bikeshed this directory structure and naming.
> func Backup(
> ```

I think the call in cli/backup_experimental.go needs to be updated, too. (The build tag situation on that file is dumb and doesn't catch things like this, which is entirely my fault, but I'll send a PR out to fix it.)

---

*[sql/backup.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-KQ0z-SsX4vo6vQSPr1A:-KQ0z-SsX4vo6vQSPr1B:bsndf9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d71e3438212c1fbb9143a580fbf20bf8cef243c/sql/backup.go#L88)):*
> ```Go
> // TODO(dan): Bikeshed this directory structure and naming.
> func Backup(
> 	ctx context.Context, db client.DB, base string, ts hlc.Timestamp,
> ```

This is the end timestamp of the range we're backing up, can we call it endTime?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-:-KQ0ywBo-JGJE5B2LLRd:b-qidl1z)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Haven't tried this with the reg cluster data yet, no.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/backup.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-KQ0zYdMdXo5CdK-G3bB:-KQ1oB0-vqb0MdtHxRrS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d71e3438212c1fbb9143a580fbf20bf8cef243c/sql/backup.go#L87)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> I think the call in cli/backup_experimental.go needs to be updated, too. (The build tag situation on that file is dumb and doesn't catch things like this, which is entirely my fault, but I'll send a PR out to fix it.)
</details>
Done.

---

*[sql/backup.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-KQ0z-SsX4vo6vQSPr1A:-KQ1oQRFyPymwVCIytKt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d71e3438212c1fbb9143a580fbf20bf8cef243c/sql/backup.go#L88)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> This is the end timestamp of the range we're backing up, can we call it endTime?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-:-KQ1sFkGoBqWtkGujM9B:bzfdl9i)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8819#-:-KQ27f8hlcOJmVNF6k8b:bnfp4nl)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: use timestamps for consistent, retryable multi-txn backups

Using logic similar to AS OF SYSTEM TIME, allow the creation of multiple
transactions that will remain consistent because their timestamps cannot
advance. Enable retries of each operation. In the future the fetch loop
could be made to work in parallel, but would need some mechanism to
limit the number of reads at once.
